napari/utils/_proxies.py

```diff
@@ -0,0 +1,68 @@
+import re
+from typing import Callable, Generic, TypeVar
+
+import wrapt
+
+from ..utils.translations import trans
+
+_T = TypeVar("_T")
+
+
+class ReadOnlyWrapper(wrapt.ObjectProxy):
+    """
+    Disable item and attribute setting with the exception of  ``__wrapped__``.
+    """
+
+    def __setattr__(self, name, val):
+        if name != '__wrapped__':
+            raise TypeError(
+                trans._(
+                    'cannot set attribute {name}',
+                    deferred=True,
+                    name=name,
+                )
+            )
+
+        super().__setattr__(name, val)
+
+    def __setitem__(self, name, val):
+        raise TypeError(
+            trans._('cannot set item {name}', deferred=True, name=name)
+        )
+
+
+_SUNDER = re.compile('^_[^_]')
+
+
+class PublicOnlyProxy(wrapt.ObjectProxy, Generic[_T]):
```

Comment:
This will work recursively right? (i.e. will apply not just to viewer, but also everything inside it, viewer.layers[0] for example). If so can you mention recursive nature in the doc string.

Reply:
it works recursively for _attribute_ access, but not `__getitem__`... do you want me to add that too?

Reply:
> it works recursively for attribute access, but not __getitem__... do you want me to add that too (I'm good with it)

yeah might as well i think - i see no reason no too

Reply:
added